src/relax/transform/inline_functions.cc

```diff
@@ -0,0 +1,306 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <tvm/driver/driver_api.h>
+#include <tvm/ir/function.h>
+#include <tvm/relax/analysis.h>
+#include <tvm/relax/expr.h>
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/transform.h>
+#include <tvm/relax/type.h>
+#include <tvm/tir/op.h>
+
+#include <tuple>
+#include <utility>
+
+#include "../../support/ordered_set.h"
+#include "utils.h"
+
+namespace tvm {
+namespace relax {
+
+namespace {
+
+using GlobalVarMap =
+    std::unordered_map<GlobalVar, std::unordered_set<GlobalVar, ObjectPtrHash, ObjectPtrEqual>,
+                       ObjectPtrHash, ObjectPtrEqual>;
+
+/*! \brief Utility to collect the map of callee to callers
+ *
+ * Used to identify functions that require recursion, either directly
+ * or indirectly.
+ */
+class CallMapCollector : ExprVisitor {
+ public:
+  static GlobalVarMap Collect(const IRModule& mod) {
+    CallMapCollector visitor;
+    for (const auto& [gvar, base_func] : mod->functions) {
+      if (auto func = base_func.as<Function>()) {
+        visitor.current_caller_ = gvar;
+        visitor(func.value());
+        visitor.current_caller_ = NullOpt;
+      }
+    }
+    return visitor.callee_to_caller_;
+  }
+
+ private:
+  void VisitExpr_(const CallNode* op) override {
+    ExprVisitor::VisitExpr_(op);
+
+    if (auto callee = op->op.as<GlobalVar>()) {
+      callee_to_caller_[callee.value()].insert(current_caller_.value());
+    }
+  }
+
+  Optional<GlobalVar> current_caller_ = NullOpt;
+  GlobalVarMap callee_to_caller_;
+};
+
+std::unordered_set<GlobalVar, ObjectPtrHash, ObjectPtrEqual> FindRecursiveFunctions(
```

Comment:
https://github.com/apache/tvm/blob/unity/src/relax/analysis/detect_recursion.cc Not sure if my analysis might be suitable for this purpose.

Reply:
Ooh, thank you for pointing that out, and that's exactly what I was looking for.  Updated the PR to use `DetectRecursion`.